src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,11 +1,106 @@
 package com.es.phoneshop.web;
 
+import com.es.phoneshop.dao.ArrayListProductDao;
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.exception.ProductNotFoundException;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartService;
+import com.es.phoneshop.model.cart.DefaultCartService;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.viewhistory.DefaultViewHistoryService;
+import com.es.phoneshop.model.viewhistory.ViewHistory;
+import com.es.phoneshop.model.viewhistory.ViewHistoryService;
+
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.Locale;
+
 public class ProductDetailsPageServlet extends AbstractProductServlet {
 
     protected static final String PRODUCT_DETAILS_JSP = "/WEB-INF/pages/productDetails.jsp";
+    private CartService cartService;
+    private ViewHistoryService viewHistoryService;
+    private ProductDao productDao;
 
     public ProductDetailsPageServlet() {
         super(PRODUCT_DETAILS_JSP);
     }
 
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+        cartService = DefaultCartService.getInstance();
+        viewHistoryService = DefaultViewHistoryService.getInstance();
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String productInfo = "";
+        Product product;
+        try {
+            productInfo = request.getPathInfo().substring(1);
+            Long id = Long.valueOf(productInfo);
+            product = productDao.getProduct(id);
+            request.setAttribute("product", product);
+            Cart cart = cartService.getCart(request.getSession());
+            request.setAttribute("cart", cart);
+            viewHistoryService.addProductToViewHistory(request.getSession(), product);
+            ViewHistory viewHistory = viewHistoryService.getViewHistory(request.getSession());
+            request.setAttribute("viewHistory", viewHistory);
+            request.getRequestDispatcher(PRODUCT_DETAILS_JSP).forward(request, response);
+        } catch (ProductNotFoundException | NumberFormatException ex) {
+            request.setAttribute("message", "Product " + productInfo + " not found");
+            response.sendError(404);
+        }
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```

Comment:
Try to simplify this method. Move some parts to separate private methods, also think over all possible scenarios. For example, now, if productId is incorrect, the exception will be thrown inside cartService.add and not caught anywhere.

Reply:
Don't stick to invocation of doGet from doPost method, it might be not the best idea to build this implementation

Reply:
For GET you most likely want to:
- Check if such product exists and send an error otherwise
- Add product, cart and viewHistory to request
- Forward to jsp

For POST it's about:
- Check if such product exists and send an error otherwise
- Add product, cart and viewHistory to request
- Check if quantity is correct, otherwise forward to jsp
- Add to cart, if out of stock - forward to jsp
- Redirect with success message

Not you only need to simplify the code as much as you can and write it without code duplication